Sources/SwiftSyntaxMacroExpansion/MacroSystem.swift

```diff
@@ -1204,10 +1218,33 @@ extension MacroApplication {
 
 // MARK: Freestanding macro expansion
 
+protocol MacroExpansion<ResultType> {
+  associatedtype ResultType
+  func map<T>(_ transform: (ResultType) throws -> T) rethrows -> T
+}
+
 extension MacroApplication {
+  struct StackMacroExpansion<ResultType>: MacroExpansion {
+    let expandedNode: ResultType
+    unowned let macroApplication: MacroApplication
+
+    init(expandedNode: ResultType, macro: any Macro.Type, macroApplication: MacroApplication) {
+      self.expandedNode = expandedNode
+      self.macroApplication = macroApplication
+      macroApplication.expandingFreestandingMacros.append(macro)
+    }
+
+    func map<T>(_ transform: (ResultType) throws -> T) rethrows -> T {
```

Comment:
I think `map` is a little misleading because I’d expect `map` to be side-effect free. It’s more a `takeResult` that doesn’t have a return value. With that name, it would also be good to have an assertion and ensure that  we don’t take the result twice (and thus pop from `macroApplication.expandingFreestandingMacros` twice).

Also, I think it would be worth adding a doc comment why we need to handle the result in a closure, ie. explain about how we detect circular macro expansions.

Reply:
> I think `map` is a little misleading because I’d expect `map` to be side-effect free. It’s more a `takeResult` that doesn’t have a return value. 

Actually there's at least one call site that returns a value ;) I'm typing this comment on a phone and have no easy access to the source, but I vaguely remember that's 'expandExpr'? 

Btw should we rename the method to 'consume'?

> With that name, it would also be good to have an assertion and ensure that  we don’t take the result twice (and thus pop from `macroApplication.expandingFreestandingMacros` twice).

I had this in my mind and wanted to implement it with lifetime specifiers. I forgot to fallback to assertions after I noticed 'consuming' is not a thing in earlier Swift versions.

> Also, I think it would be worth adding a doc comment why we need to handle the result in a closure, ie. explain about how we detect circular macro expansions.

True!

Reply:
> I had this in my mind and wanted to implement it with lifetime specifiers. I forgot to fallback to assertions after I noticed 'consuming' is not a thing in earlier Swift versions.

> Sans the support of move-only value types, we need to change StackMacroExpansion from a struct to a class too.

Yeah, we can’t use those yet and even then, I think it might be tricky to keep the move-only type alive until all the nested macro expansion are performed. The current design is probably best for now.